File: vol1/ch11/LeetCode2536-1.cpp
class Solution {
public:
    vector<vector<int>> rangeAddQueries(int n,vector<vector<int>>&queries) {
        vector<vector<int>> diff(n,vector<int>(n,0));
        for(int p=0;p<queries.size();p++) {
            int r1=queries[p][0],c1=queries[p][1];
            int r2=queries[p][2],c2=queries[p][3];
            for(int i=r1;i<=r2;i++) {
                diff[i][c1]+=1;
                if(c2+1<n) diff[i][c2+1]-=1;
            }
        }
        vector<vector<int>> ans(n,vector<int>(n,0)); 
        for(int i=0;i<n;i++)
            ans[i][0]=diff[i][0];
        for(int i=0;i<n;i++) {
            for(int j=1;j<n;j++)
                ans[i][j]=ans[i][j-1]+diff[i][j];
        }
        return ans;
    }
};

File: vol1/ch11/LeetCode2536-1.py
class Solution:
    def rangeAddQueries(self, n: int, queries: List[List[int]]) -> List[List[int]]:
        diff=[[0]*n for i in range(n)]
        for p in range(0,len(queries)):
            r1,c1=queries[p][0],queries[p][1]
            r2,c2=queries[p][2],queries[p][3]
            for i in range(r1,r2+1):
                diff[i][c1]+=1
                if c2+1<n:diff[i][c2+1]-=1
        ans=[[0]*n for i in range(n)]
        for i in range(0,n):
            ans[i][0]=diff[i][0]
        for i in range(0,n):
            for j in range(1,n):
                ans[i][j]=ans[i][j-1]+diff[i][j]
        return ans
